### ISSUE TITLE
Module fails validation in dxc but passes in dxil

### ISSUE THREAD

**Issue by seanbaxter** [microsoft/DirectXShaderCompiler#3440]
https://gist.github.com/seanbaxter/27b35a616f1de0c2ffff1641695936ba
The dxil.ll here fails validation when called on the DxilModule using dxc's validator.
> error: Function: frag_main: error: Resource handle should returned by createHandle. Use /Zi for source location.

The error is thrown from GetResourceFromHandle called by ValidateSampleInst, when GetResourceKindAndCompTy checks the srvHandle ResourceKind. The DxilResourceProperties structure is all zeroes, and the Invalid Basic.ResourceKind triggers the exception.

The preceding GetResourceFromHandle that checks the samplerHandler does yield a non-zero resource props, that indicates the sampler resource kind.

Because this is ps_6_5 code, there are no actual resource props in the metadata, so whatever code is creating them in the validator is messing up. Perhaps the DxilModule is out of sync from a legalization maneuver. I inserted `dxil_module.GetOP()->RefreshCache()` right prior to ValidateDxilModule, but that doesn't fix it.

So, I had to turn validation off in the compiler to generate a module. Would like to know workaround to turn it back on.

**Comment by pow2clk** [comment]
Hi sean,

Can you clarify a few things? The title says it "passes in dxil".  What do you mean by that? The description doesn't mention anything passing. Are you saying the dxil is able to be consumed by a hardware driver?

What "dxc validator" are you referring to? If it's dxv.exe, I was able to validate the .ll file you referenced successfully. 

From context, I think that `GetResourceHandle` is calling `GetResourceFromVal` and failing to find an entry for the given handle in  `ResPropMap`, which is populated for all shader models by `BuildResMap` based on either `createHandle` or `AnnotateHandle` depending. If you're calling these functions from your own code, that is where the resource would have to succeed or fail to get its resource props assigned.

**Comment by seanbaxter** [comment]
I wrote a dxil-signing.exe that opens dxil.dll and does validation and signing. That part passes. The thing that failed was ValidateDxilModule. But you pointed me to `BuildResMap` and I was able to easily back it out from there and fix my issue. 

Basically, I wasn't calling DxilResource::SetRW, so the ResourceClass for my buffers was still Invalid (as set by DxilResource's ctor), causing DxilResourceProperties.cpp:178 to return a zeroed-out RP. However I was writing a proper DxilResourceBinding, which is why the containerized module was correct. I didn't know this second point of definition, but your comment showed me it.

Glad to have ValidateDxilModule back online. Thanks Greg.
